torchserve-vicuna/langchain_demo_vicuna_torchserve_grpc.py

```diff
@@ -0,0 +1,156 @@
+        return "custom"
+
+
+llm = CustomLLM()
+
+memory = ConversationBufferWindowMemory(k=3, memory_key="chat_history")
+
+template = "{chat_history} Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\n{question}\n\n### Response:\n"
+
+prompt = PromptTemplate(template=template, input_variables=["chat_history", "question"])
+
+llm_chain = LLMChain(prompt=prompt, llm=llm, memory=memory, output_key="result")
+
+
+def run_query(question):
+    global responses
+    llm_chain.run(question)
+
+class Seafoam(Base):
+    def __init__(
+        self,
+        *,
+        primary_hue: colors.Color | str = colors.emerald,
+        secondary_hue: colors.Color | str = colors.blue,
+        neutral_hue: colors.Color | str = colors.gray,
+        spacing_size: sizes.Size | str = sizes.spacing_md,
+        radius_size: sizes.Size | str = sizes.radius_md,
+        text_size: sizes.Size | str = sizes.text_lg,
+        font: fonts.Font | str | Iterable[fonts.Font | str] = (
+            fonts.GoogleFont("Quicksand"),
+            "ui-sans-serif",
+            "sans-serif",
+        ), font_mono: fonts.Font | str | Iterable[fonts.Font | str] = (
+            fonts.GoogleFont("IBM Plex Mono"),
+            "ui-monospace",
+            "monospace",
+        ),
+    ):
+        super().__init__(
+            primary_hue=primary_hue,
+            secondary_hue=secondary_hue,
+            neutral_hue=neutral_hue,
+            spacing_size=spacing_size,
+            radius_size=radius_size,
+            text_size=text_size,
+            font=font,
+            font_mono=font_mono,
+        )
+
+def launch_gradio_interface():
+    CSS ="""
+    .contain { display: flex; flex-direction: column; }
+    #component-0 { height: 100%; }
+    #chatbot { flex-grow: 1; }
+    #text_box.gradio-container { background-color: transparent; }
+    #send_button { background-color: #6ee7b7; margin-top: 2.5%}
+    """
+
+    seafoam = Seafoam()
+    def user(user_message, history):
+        return gr.update(value="", interactive=False), history + [[user_message, None]]
+
+    def bot(history):
+        global responses
+        print("Sending Query!", history[-1][0])
+        run_query(history[-1][0])
+        history[-1][1] = ""
+        try:
+            flag = False
+            for resp in responses:
+                prediction = resp.prediction.decode("utf-8")
+                print(prediction, flush=True, end="")
+                if "### Response:" in prediction:
+                    flag = True
+                    continue 
+                if flag:  
+                    history[-1][1] += prediction
+                    time.sleep(0.05)
+                    yield history
+            flag = False
+        except grpc.RpcError as e:
+            exit(1)
+
+    with gr.Blocks(css=CSS, theme=seafoam) as demo:
+        chatbot = gr.Chatbot(label="PyTorch Bot", show_label=True, elem_id="chatbot")
```

Comment:
The gradio chatbot already taking care of chat history handling, why do we need to manually add the user / bot history.